### PR TITLE
Added drag and drop feature to the code

### PR DESCRIPTION
**Pull request by jfmherokiller**

### CODE REVIEW

File: src/citra_qt/game_list.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@ GameList::GameList(QWidget* parent) : QWidget{parent} {
     item_model->setHeaderData(COLUMN_NAME, Qt::Horizontal, "Name");
     item_model->setHeaderData(COLUMN_FILE_TYPE, Qt::Horizontal, "File type");
     item_model->setHeaderData(COLUMN_SIZE, Qt::Horizontal, "Size");
-
+    this->setAcceptDrops(true);
```
Comment (yuriks) [inline src/citra_qt/game_list.cpp]: Don't use `this->` unless you need to

File: src/citra_qt/game_list.cpp [hunk 2]
```diff
@@ -185,3 +185,17 @@ void GameListWorker::Cancel() {
...
+    const QMimeData* mimeData = event->mimeData();
+    QString filename = mimeData->urls().at(0).toLocalFile();
+    if (HasSupportedFileExtension(filename.toStdString())) {
+        emit GameChosen(filename);
+    }
+}
+void GameList::dragEnterEvent(QDragEnterEvent* event) {
+    event->acceptProposedAction();
+}
+void GameList::dragMoveEvent(QDragMoveEvent* event) {
```
Comment (jfmherokiller) [inline src/citra_qt/game_list.cpp]: yes because without it the file isnt accepted at all

File: src/citra_qt/game_list.cpp [hunk 3]
```diff
@@ -185,3 +185,17 @@ void GameListWorker::Cancel() {
...
+
+void GameList::dropEvent(QDropEvent* event) {
+    const QMimeData* mimeData = event->mimeData();
+    QString filename = mimeData->urls().at(0).toLocalFile();
+    if (HasSupportedFileExtension(filename.toStdString())) {
+        emit GameChosen(filename);
+    }
+}
+void GameList::dragEnterEvent(QDragEnterEvent* event) {
+    event->acceptProposedAction();
```
Comment (jfmherokiller) [inline src/citra_qt/game_list.cpp]:
```
void GameList::dragEnterEvent(QDragEnterEvent* event) { 
  const QMimeData* mimeData = event->mimeData();
  QString filename = mimeData->urls().at(0).toLocalFile();
  if (HasSupportedFileExtension(filename.toStdString())) {
    event->acceptProposedAction();
  }
}
```
would this be better?

File: src/citra_qt/game_list.cpp [hunk 4]
```diff
@@ -185,3 +185,17 @@ void GameListWorker::Cancel() {
     disconnect(this, nullptr, nullptr, nullptr);
     stop_processing = true;
 }
+
+void GameList::dropEvent(QDropEvent* event) {
+    const QMimeData* mimeData = event->mimeData();
+    QString filename = mimeData->urls().at(0).toLocalFile();
```
Comment (jfmherokiller) [inline src/citra_qt/game_list.cpp]: well should i put up an error message or deny the drop event if there exists more than one file or if the file is not valid?

Comment (yuriks) [inline src/citra_qt/game_list.cpp]: Just deny/don't accept the event. That'll cause the OS to show an "invalid" icon and reject drop attempt.

File: src/citra_qt/game_list.cpp (hunk 3 above)
Comment (yuriks) [inline src/citra_qt/game_list.cpp]: `if (mimeData->hasUrls() && mimeData->urls().length() == 1)`

File: src/citra_qt/game_list.cpp (hunk 2 above)
Comment (yuriks) [inline src/citra_qt/game_list.cpp]: Ok.

File: src/citra_qt/main.cpp [hunk 5]
```diff
@@ -654,3 +654,29 @@ int main(int argc, char* argv[]) {
     main_window.show();
     return app.exec();
 }
+void GMainWindow::dropEvent(QDropEvent* event) {
+    if (ConfirmChangeGame()) {
+        const QMimeData* mimeData = event->mimeData();
+        QString filename = mimeData->urls().at(0).toLocalFile();
```
Comment (yuriks) [inline src/citra_qt/main.cpp]: You should still do the check to see if it's a valid event here. (Preferably factor that out into a small function so it's not duplicated between dragEventEnter and dropEvent.)

File: src/citra_qt/main.cpp [hunk 6]
```diff
@@ -654,3 +654,29 @@ int main(int argc, char* argv[]) {
...
+        BootGame(filename.toStdString());
+    }
+}
+void GMainWindow::dragEnterEvent(QDragEnterEvent* event) {
+    const QMimeData* mimeData = event->mimeData();
+    if (mimeData->hasUrls() && (mimeData->urls().length() == 1)) {
```
Comment (yuriks) [inline src/citra_qt/main.cpp]: You don't need the extra parenthesis around the second part of the condition.

File: src/citra_qt/main.cpp [hunk 7]
```diff
@@ -654,3 +654,32 @@ int main(int argc, char* argv[]) {
     main_window.show();
     return app.exec();
 }
+bool checkeventvalid(QDropEvent* event) {
```
Comment (yuriks) [inline src/citra_qt/main.cpp]: Name this something more descriptive, like `IsSingleFileDropEvent`.

File: src/citra_qt/main.cpp [hunk 8]
```diff
@@ -654,3 +654,32 @@ int main(int argc, char* argv[]) {
     main_window.show();
     return app.exec();
 }
```
Comment (yuriks) [inline src/citra_qt/main.cpp]: You should leave a blank line between each function definition. Same for all the other functions you added.

File: src/citra_qt/main.cpp (hunk 7 above)
Comment (yuriks) [inline src/citra_qt/main.cpp]: Also, move your function above the the `main` function (and the ifdefs above it) so it stays with the rest of the GMainWindow code.

File: src/citra_qt/main.cpp [hunk 9]
```diff
@@ -624,6 +624,39 @@ void GMainWindow::closeEvent(QCloseEvent* event) {
 
     QWidget::closeEvent(event);
 }
+bool IsSingleFileDropEvent(QDropEvent* event) {
```
Comment (yuriks) [inline src/citra_qt/main.cpp]: You forgot a newline here...

File: src/citra_qt/main.cpp [hunk 10]
```diff
@@ -624,6 +624,39 @@ void GMainWindow::closeEvent(QCloseEvent* event) {
...
+    if (IsSingleFileDropEvent(event)) {
+        event->acceptProposedAction();
+    }
+}
+
+void GMainWindow::dragMoveEvent(QDragMoveEvent* event) {
+    event->acceptProposedAction();
+}
+
+bool GMainWindow::ConfirmChangeGame() {
```
Comment (jroweboy) [inline src/citra_qt/main.cpp]: Maybe in a follow up PR, you could reuse ConfirmChangeGame to ask when they click "Open File" while a game is running if they want to change games

File: src/citra_qt/main.cpp (hunk 9 above)
Comment (jfmherokiller) [inline src/citra_qt/main.cpp]: oh cmon lol

File: src/citra_qt/main.cpp (hunk 10 above)
Comment (jfmherokiller) [inline src/citra_qt/main.cpp]: not a bad idea ill see if i can do it

Comment (jfmherokiller) [inline src/citra_qt/main.cpp]: @jroweboy ill begin working on that pull request idea now

**Pull request merged**